### ISSUE TITLE
JAX-RS class hierarchy scanning is incomplete and creates transaction name explosion when using EJB

### ISSUE THREAD

**Issue by mdindoffer** [elastic/apm-agent-java#512]
**Summary**
The JAX-RS class hierarchy scanning introduced in PR #493   does not work correctly when using EJB.
EJB uses a `$$$view<number>` naming scheme when creating proxies, which do not get filtered out and leads to explosion of transaction names:
 
![screenshot from 2019-03-04 15-18-38_crop](https://user-images.githubusercontent.com/12681399/53796890-5990f700-3f35-11e9-85c9-60a1c90980cc.png)
![screenshot from 2019-03-04 15-18-10_crop_blurred](https://user-images.githubusercontent.com/12681399/53800135-52b9b280-3f3c-11e9-88db-636beb1902cc.png)

**Additional Info**
I've tried to fix it by simply adding a condition to the `JaxRsTransactionNameInstrumentation` matcher:
```java
.and(not(ElementMatchers.<TypeDescription>nameContains("$view")))
```
That gets rid of the `$$$view` explosion, but all the requests still end up in the HttpServlet30Dispatcher transaction:
![screenshot_2019-03-05 httpservlet30dispatcher_dopost_blurred](https://user-images.githubusercontent.com/12681399/53800839-f22b7500-3f3d-11e9-9c4a-506bdf0b4fa7.png)
This shows that the current solution from PR elastic/apm-agent-java#493 is not sufficient. The problem probably comes up when dealing with generic hierarchies.

**A Reproducible scenario**
We use a generic structure that might look complicated and intimidating at first glance, but that's just because of the number of generic types used. We have an abstract generic endpoint, a concrete endpoint specialization, a generic endpoint/client interface and a concrete endpoint/client interface.
To illustrate (Wildfly 9.0.2 btw):
1. Concrete JaxRS Endpoint:
```java
import javax.ejb.Singleton;

@Singleton
public class ConcreteEndpoint
        extends AbstractRequestEndpoint<ClassT, ClassU, ClassE, ClassV>
        implements ConcreteClient
```
2. Concrete endpoint/client JaxRS web service interface definition:
```java
@Path( "/" )
public interface ConcreteClient extends BaseRequestClient<ClassT, ClassU>
```
3. Abstract JaxRS endpoint:
```java
public abstract class AbstractRequestEndpoint<T extends Class1<?>, U extends Class2<?>, E extends Class3, V extends Class4<E, U>>
        implements BaseRequestClient<T, U>
```
4. Generic JaxRS web service interface definition:
```java
public interface BaseRequestClient<T extends Class1<?>, U extends Class2<?>>
```
**Expected results**
We need to scan the complete jaxrs class hierarchy and find the `@Path` annotations, but ignore the `view` ejb proxies.

Maybe there are two separate issues, but I feel they are tightly coupled, so I've created just this one issue.

**Comment by eyalkoren** [comment]
Looks like the proxy class names start with `view`. Did you try `.and(not(ElementMatchers.<TypeDescription>nameStartsWith("view")))`?
If that works we can enhance to ignore classes with the general structure `view*#` to make it more specific

**Comment by felixbarny** [comment]
I think we should consider classes containing `$$$` proxy classes and ignore them generally.

**Comment by mdindoffer** [comment]
Ok, so now I finally know what's going on. Leaving aside filtering of proxies (might filter by $$$ or by another pattern), there is a problem with class hierarchy scanning. Fixing the `typeMatcher` condition correctly resolves the endpoint classes and applies the advice:
```
DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Type match for advice JaxRsTransactionNameInstrumentation: (((not(isInterface()) and not(name(contains($Proxy)))) and not(name(contains($view)))) and (declaresAnnotations(whereOne(ofAnnotationType(name(equals(javax.ws.rs.Path))))) or hasSuperType(erasure(declaresAnnotations(whereOne(ofAnnotationType(name(equals(javax.ws.rs.Path))))))))) matches class com.example.ConcreteEndpoint
```

In our use case we have a generic endpoint that's represented by an `abstract` java class. The concrete endpoint class is just a generic specialization of the abstract endpoint, thus does not `@Override` most of the REST endpoint methods from the superclass. In fact, it might simply be an empty class:
```java
public class ConcreteEndpoint extends AbstractEndpoint<T> implements ConcreteClient
{
//NOOP
}
```

When such class is matched by a `typeMatcher` and visited by the ASM class visitor, only the methods explicitly declared in the class are visited and checked against the `methodMatcher`. Since such class does not declare any methods, `methodMatcher` has nothing to match against.
It seems that the ASM class visitor visits only one class and does not walk the hierarchy.
Is there a way to tell ASM to visit methods of parent class?

**Comment by felixbarny** [comment]
Hmm, that patter should actually work. I've created a test for this scenario here: https://github.com/elastic/apm-agent-java/pull/517

Could it be that the `AbstractEndpoint` does not have any JAX-RS annotations?

**Comment by mdindoffer** [comment]
> Could it be that the `AbstractEndpoint` does not have any JAX-RS annotations?

Yes. Here's a pic to make it clearer:
![diagram](https://user-images.githubusercontent.com/12681399/54113718-2fce4900-43e9-11e9-8503-840035280e97.png)

**Comment by felixbarny** [comment]
I see. But I'm not sure there's a proper solution for this. Why don't you just put the JAX-RS annotations on the `AbstractEndpoint` or the `AbstractClient`? Do you know that annotations are not required to be on the class path? So if you would use `AbstractEndpoint` or `AbstractClient` in an application which has no dependency on JAX-RS, you wouldn't see `ClassNotFoundExceptions`. That may also simplify your class hierarchy.

**Comment by mdindoffer** [comment]
Because the Concrete Endpoints listen on different paths.
We have 5 `ConcreteClient` interfaces and each one has different `@Path` annotations. How would you pull them up?

Are you suggesting I should put a "dummy" path annotation on the `AbstractClient` and override it downstream with proper paths? That may be a solution, but I was hoping for a somewhat cleaner fix.

**Comment by mdindoffer** [comment]
> Are you suggesting I should put a "dummy" path annotation on the `AbstractClient` and override it downstream with proper paths? That may be a solution, but I was hoping for a somewhat cleaner fix.

Actually, this does not work at all. The `AbstractClient` is generic (that's the whole point of the hierarchy). When I tried to annotate a generic resource with `@Path` JAX-RS annotation Wildfly deployment failed with (Yeah I know Resteasy isn't officially supported):
```
2019-03-11 12:36:57,771 ERROR [org.jboss.msc.service.fail] (ServerService Thread Pool -- 7) MSC000001: Failed to start service jboss.undertow.deployment.default-server.default-host./api/v1/example-endpoint: org.jboss.msc.service.StartException in service jboss.undertow.deployment.default-server.default-host./api/v1/example-endpoint: java.lang.RuntimeException: RESTEASY003970: Unable to resolve type variable
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentService$1.run(UndertowDeploymentService.java:85)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:320)
Caused by: java.lang.RuntimeException: RESTEASY003970: Unable to resolve type variable
	at org.jboss.resteasy.util.Types.getImplementingMethod(Types.java:163)
	at org.jboss.resteasy.spi.metadata.ResourceBuilder.findAnnotatedInterfaceMethod(ResourceBuilder.java:779)
	at org.jboss.resteasy.spi.metadata.ResourceBuilder.findAnnotatedMethod(ResourceBuilder.java:840)
	at org.jboss.resteasy.spi.metadata.ResourceBuilder.processMethod(ResourceBuilder.java:910)
	at org.jboss.resteasy.spi.metadata.ResourceBuilder.fromAnnotations(ResourceBuilder.java:759)
	at org.jboss.resteasy.spi.metadata.ResourceBuilder.rootResourceFromAnnotations(ResourceBuilder.java:709)
	at org.jboss.resteasy.plugins.server.resourcefactory.POJOResourceFactory.<init>(POJOResourceFactory.java:30)
	at org.jboss.resteasy.core.ResourceMethodRegistry.addPerRequestResource(ResourceMethodRegistry.java:76)
	at org.jboss.resteasy.spi.ResteasyDeployment.registration(ResteasyDeployment.java:409)
	at org.jboss.resteasy.spi.ResteasyDeployment.start(ResteasyDeployment.java:250)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.init(ServletContainerDispatcher.java:113)
	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.init(HttpServletDispatcher.java:36)
	at io.undertow.servlet.core.LifecyleInterceptorInvocation.proceed(LifecyleInterceptorInvocation.java:117)
	at org.wildfly.extension.undertow.security.RunAsLifecycleInterceptor.init(RunAsLifecycleInterceptor.java:78)
	at io.undertow.servlet.core.LifecyleInterceptorInvocation.proceed(LifecyleInterceptorInvocation.java:103)
	at io.undertow.servlet.core.ManagedServlet$DefaultInstanceStrategy.start(ManagedServlet.java:230)
	at io.undertow.servlet.core.ManagedServlet.createServlet(ManagedServlet.java:131)
	at io.undertow.servlet.core.DeploymentManagerImpl.start(DeploymentManagerImpl.java:511)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentService.startContext(UndertowDeploymentService.java:101)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentService$1.run(UndertowDeploymentService.java:82)
	... 7 more
```

It seems to me our design is clean and OOP-OK (and works). It does not feel right to change our architecture in a major way just to conform with an APM agent, so I'd like to find a way to fix it in the agent.

**Comment by felixbarny** [comment]
> Because the Concrete Endpoints listen on different paths.

Gotcha. The problem is that the transaction name is set in the method handling the request - in this case `AbstractEndpoint`. To avoid overhead, the hierarchy is not determined at runtime. The name basically gets hardcoded into the method via instrumentation as the class is loaded. Therefore, the name can't be determined based on the runtime type of `ConcreteClient` and its interfaces.

One solution would be to determine the name at runtime. But there is still the challenge that we can't know that `AbstractEndpoint` is a JAX-RS resource just by looking at the class by itself. 

Another approach would be to generate overriding methods in `ConcreteEndpoint` via the agent. But that means we can't attach the agent at runtime because adding methods is not allowed when redefining classes. That approach would also be quite intrusive.

Yet another idea would be to instrument the concrete JAX-RS implementation and hook into the part where the request is dispatched to a concrete resource class. Based on that we should be able to give the transaction a proper name. However, that would be specific to a implementation and would most likely be quite brittle as it relies on internals of the implementation which could change even within the same minor version.

**Comment by eyalkoren** [comment]
Maybe we miss them now because the `avax.ws.rs.GET/POST/...` annotations are used in the Proxy classes that you excluded from instrumentation? If that's the case, the solution would be to include them in the instrumentation but not use their class name, but the actual class name they are derived from.

**Comment by felixbarny** [comment]
Another idea: I think RestEasy provides some kind of callbacks before a resource is invoked, right? Could you use those in combination with the [`Transaction#setName` api](https://www.elastic.co/guide/en/apm/agent/java/current/public-api.html#api-set-name) to manually override the transaction name?

**Comment by mdindoffer** [comment]
You're right! I didn't know about this, but we can use the [ContainerRequestFilter](https://docs.oracle.com/javaee/7/api/javax/ws/rs/container/ContainerRequestFilter.html) from Java EE spec and then cast the `ContainerRequestContext` to RestEasy's `PostMatchContainerRequestContext` that has access to the resolved resource method. See the [example filter](https://gist.github.com/mdindoffer/bc7cf99b5384f7dcbd85db72ad0de192)

A prerequisite for this to work is to have the `$view` filtered out of the `JaxRsTransactionNameInstrumentation`, otherwise the transaction name gets overwritten with the proxy class name.

The question is, do I like this solution? Not really. It's a hack relying on both a JavaEE container being present together with RestEasy, instead of simple JAX-RS.

Also, this results in a small inconsistency in Tx naming (although that's nitpicking):
Other web resources that are not part of the generic interface hierarchy end up being named by their implementation class name (e.g. `SomeEndpoint#method`), while these resources from the hierarchy end up being named with the client interface name (e.g. `ConcreteClient#method`).

Anyway thanks for the suggestion. I think we can live with the workaround for now and I don't want to block the agent's release.

**Comment by felixbarny** [comment]
Closed via #600
